### PR TITLE
KAFKA-10338; Support PEM format for SSL key and trust stores (KIP-651)

### PR DESCRIPTION
**Pull request by rajinisivaram**
Adds support for SSL key and trust stores to be specified in PEM format either as files or directly as configuration values.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by rajinisivaram** [comment]
retest this please

File: clients/src/test/java/org/apache/kafka/common/network/CertStores.java [hunk 1]
```diff
@@ -54,13 +60,30 @@ public CertStores(boolean server, String commonName, InetAddress hostAddress) th
...
+                .usePem(usePem)
+                .build();
     }
 
+
     public Map<String, Object> getTrustingConfig(CertStores truststoreConfig) {
+        return getTrustingConfig(truststoreConfig, false);
+    }
+
+    public Map<String, Object> getTrustingConfig(CertStores truststoreConfig, boolean usePemCerts) {
```
Comment (omkreddy) [inline clients/src/test/java/org/apache/kafka/common/network/CertStores.java]: unused usePemCerts?

File: clients/src/test/java/org/apache/kafka/test/TestSslUtils.java [hunk 2]
```diff
@@ -199,6 +206,156 @@ public static void createKeyStore(String filename,
...
+        if (certChain == null && ksPath != null) {
+            String ksType = (String) sslProps.get(SslConfigs.SSL_KEYSTORE_TYPE_CONFIG);
+            Password ksPassword = (Password) sslProps.remove(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG);
+            Password keyPassword = (Password) sslProps.get(SslConfigs.SSL_KEY_PASSWORD_CONFIG);
+            certChain = exportCertificates(ksPath, ksPassword, ksType);
+            Password pemKeyPassword = encryptPrivateKey ? keyPassword : null;
+            key = exportPrivateKey(ksPath, ksPassword, keyPassword, ksType, pemKeyPassword);
+            if (!encryptPrivateKey)
+                sslProps.remove(SslConfigs.SSL_KEY_PASSWORD_CONFIG);
+        } else if (!encryptPrivateKey) {
```
Comment (omkreddy) [inline clients/src/test/java/org/apache/kafka/test/TestSslUtils.java]: Can we remove empty else block?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 3]
```diff
@@ -923,9 +926,12 @@ object KafkaConfig {
   val SslKeystoreLocationDoc = SslConfigs.SSL_KEYSTORE_LOCATION_DOC
   val SslKeystorePasswordDoc = SslConfigs.SSL_KEYSTORE_PASSWORD_DOC
   val SslKeyPasswordDoc = SslConfigs.SSL_KEY_PASSWORD_DOC
+  val SslKeystoreKeyDoc = SslConfigs.SSL_KEYSTORE_KEY_DOC
+  val SslKeystoreCertificateChainDoc = SslConfigs.SSL_KEYSTORE_CERTIFICATE_CHAIN_CONFIG
   val SslTruststoreTypeDoc = SslConfigs.SSL_TRUSTSTORE_TYPE_DOC
   val SslTruststorePasswordDoc = SslConfigs.SSL_TRUSTSTORE_PASSWORD_DOC
   val SslTruststoreLocationDoc = SslConfigs.SSL_TRUSTSTORE_LOCATION_DOC
+  val SslTruststoreCertificatesDoc = SslConfigs.SSL_TRUSTSTORE_CERTIFICATES_CONFIG
```
Comment (omkreddy) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: SslConfigs.SSL_TRUSTSTORE_CERTIFICATES_CONFIG => SslConfigs.SSL_TRUSTSTORE_CERTIFICATES_DOC

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -923,9 +926,12 @@ object KafkaConfig {
   val SslKeystoreLocationDoc = SslConfigs.SSL_KEYSTORE_LOCATION_DOC
   val SslKeystorePasswordDoc = SslConfigs.SSL_KEYSTORE_PASSWORD_DOC
   val SslKeyPasswordDoc = SslConfigs.SSL_KEY_PASSWORD_DOC
+  val SslKeystoreKeyDoc = SslConfigs.SSL_KEYSTORE_KEY_DOC
+  val SslKeystoreCertificateChainDoc = SslConfigs.SSL_KEYSTORE_CERTIFICATE_CHAIN_CONFIG
```
Comment (omkreddy) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: SslConfigs.SSL_TRUSTSTORE_CERTIFICATES_CONFIG => SslConfigs.SSL_TRUSTSTORE_CERTIFICATES_DOC

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java [hunk 5]
```diff
@@ -167,17 +181,12 @@ public void testValidEndpointIdentificationSanIp() throws Exception {
     @Test
     public void testValidEndpointIdentificationCN() throws Exception {
         String node = "0";
```
Comment (omkreddy) [inline clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: unused `node` variable here and other methods.

**Review by omkreddy — APPROVED** [review]
@rajinisivaram Thanks for the PR. LGTM.

File: clients/src/test/java/org/apache/kafka/common/network/CertStores.java (hunk 1 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/network/CertStores.java]: removed

File: clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java (hunk 5 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/network/SslTransportLayerTest.java]: fixed these and few other warnings

**Comment by rajinisivaram** [comment]
@omkreddy Thanks for the review, have addressed the comments.

**Comment by rajinisivaram** [comment]
@omkreddy Thanks for the review, merging to trunk.

File: clients/src/main/java/org/apache/kafka/common/security/ssl/DefaultSslEngineFactory.java [hunk 6]
```diff
@@ -246,56 +269,104 @@ private SSLContext createSSLContext() {
...
+                throw new InvalidConfigurationException("SSL key store password cannot be specified with PEM format, only key password may be specified.");
+            else
+                return new PemStore(certificateChain, privateKey, keyPassword);
+        } else if (certificateChain != null) {
+            throw new InvalidConfigurationException("SSL certificate chain is specified, but private key is not specified");
+        } else if (PEM_TYPE.equals(type) && path != null) {
+            if (password != null)
+                throw new InvalidConfigurationException("SSL key store password cannot be specified with PEM format, only key password may be specified");
+            else if (keyPassword == null)
+                throw new InvalidConfigurationException("SSL PEM key store is specified, but key password is not specified.");
```
Comment (amccague) [inline clients/src/main/java/org/apache/kafka/common/security/ssl/DefaultSslEngineFactory.java:291]: Is there an intention here to not allow a `FileBasedPemStore` containing an unencrypted private key when a key could be provided to `PemStore` with a `null`  `keyPassword`?

**Comment by vendamere** [comment]
Was this change meant to work for the schema registry client as well? I've tried connecting to the schema registry using these settings, but it fails with a message indicating it's looking for a file still.

**Comment by bachmanity1** [comment]
> How is PEM certificate renewal possible on the producer/consumer client? Is this documented anywhere?

Hi @teabot , could find the answer to your question? I have also faced a similar. More precisely, is it possible to update the certificate without requiring a restart of the broker?

**Pull request merged**